CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

IntegralProgrammer opened the pull request:
The PR implements CARDS-1492 by providing various performance counters in the JCR under `/PerformanceStatistics/` and a periodic task which sends updates to a Slack channel with the values of the performance counters.

To test:

Testing setup
------------------

1. Build the _CARDS-1492_ branch (`mvn clean install`)
2. If you are using the provided development _mock Slack_ server, open a new terminal and run `cd Utilities/Development`, then `nodejs mock_slack.js`. Otherwise, if you are using the private _cards-alerts-dev_ channel, please message me and I will add you to this channel and provide you with the private URL to POST to for sending messages to that channel.
3. Start CARDS in _Cards4Proms_ mode with the command `SLACK_PERFORMANCE_URL=http://localhost:8000/ PATIENT_NOTIFICATION_FROM_ADDRESS='datapro.uhn.ca' PATIENT_NOTIFICATION_FROM_NAME='UHN DATAPRO' NIGHTLY_NOTIFICATIONS_SCHEDULE='0 * * * * ? *' SLING_COMMONS_CRYPTO_PASSWORD=password CARDS_HOST_AND_PORT='localhost:8080' NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE='0 * * * * ? *' ./start_cards.sh --project cards4proms --dev`
4. Every minute, you should see a new Slack message being posted with the _Today_ and _Total_ values for the performance statistics of _Initial Emails Sent_, _Reminder Emails Sent_, _Total Surveys Submitted_, _Appointment Surveys Submitted_, and _Imported Appointments_. Please note that given, for testing purposes, we have set this task to run every minute instead of once a day, a day in this testing scenario is considered to be one minute long, therefore the _Today_ values for a statistic are reset every minute.

Testing _Initial Emails Sent_ Performance Statistic
---------------------------------------------------------------------

1. In a new terminal, start the mock Torch server for an appointment 3 days from now by running `cd Utilities/Development`, then `nodejs mock_torch.js --appointment-time-hours-from-now=72`.
2. Visit `http://localhost:8080/system/console/configMgr`.
3. Look for the `PROMs import` configuration, click on it to add a new config, set the _Name_ to `mock`, _Endpoint URL_ to `http://localhost:8011/`, and _Clinic_ to `6012-HC-Congenital Cardiac`. Save the configuration.
4. Point your browser to `http://localhost:8080/Subjects.importTorch.json?config=mock`.
5. Visit `http://localhost:8080` once again. You should see an imported patient and their scheduled visit.
6. On Slack (or the mock Slack server), a message should be posted for `Initial Emails Sent -- Today: 1, Total: 1`, then one minute later, `Initial Emails Sent -- Today: 1, Total: 2`, then one minute later, `Initial Emails Sent -- Today: 1, Total: 3`, ...etc.

Testing _Reminder Emails Sent_ Performance Statistic
--------------------------------------------------------------------------

1. Stop the `mock_torch.js` server and restart it with the parameter `--appointment-time-hours-from-now=24` (instead of `--appointment-time-hours-from-now=72`).
2. Point your browser to `http://localhost:8080/Subjects.importTorch.json?config=mock`. This import should cause the initially imported appointment to be moved from 3 days from now to 1 day from now.
3. On Slack (or the mock Slack server), a message should be posted for `Reminder Emails Sent -- Today: 1, Total: 1`, then one minute later, `Reminder Emails Sent -- Today: 1, Total: 2`, then one minute later, `Reminder Emails Sent -- Today: 1, Total: 3`, ...etc.

Testing _Total Surveys Submitted_ Performance Statistic
------------------------------------------------------------------------------

1. Visit `http://localhost:8080/system/console/configMgr`.
2. Look for the `Submission Counter` configuration, click on it to add a new config, and save the default values.
3. Open the imported _Visit information_ Form and set the _completed_ and _submitted_ values to _Yes_. Save the Form.
4. On Slack (or the mock Slack server), a message should be posted for `Total Surveys Submitted: Today: 5, Total: 5` then `Total Surveys Submitted: Today: 0, Total: 5` then `Total Surveys Submitted: Today: 0, Total: 5`, ...etc.

Testing _Appointment Surveys Submitted_ Performance Statistic
-----------------------------------------------------------------------------------------

1. On Slack (or the mock Slack server), a message should be posted for `Appointment Surveys Submitted: Today: 1, Total: 1` then `Appointment Surveys Submitted: Today: 0, Total: 1` then `Appointment Surveys Submitted: Today: 0, Total: 1`, ...etc.

Testing _Imported Appointments_ Performance Statistic
-----------------------------------------------------------------------------

1. Given that we already have imported two appointments, on Slack (or the mock Slack server), further up on the page, a message should be posted for `Imported Appointments: Today: 1, Total: 1` then `Imported Appointments: Today: 0, Total: 1` then `Imported Appointments: Today: 0, Total: 1`, ..., `Imported Appointments: Today: 1, Total: 2` then `Imported Appointments: Today: 0, Total: 2` then `Imported Appointments: Today: 0, Total: 2`, ...etc.


Please note that the _Imported Appointments_ performance counter increments each time the Torch import task runs and obtains upcoming visits and not when new visits are created or existing visits are modified. Should this behavior be changed? @marta-

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Working implementation of CARDS-1492

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Add mock Slack server for testing

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Remove debugging LOGGER calls

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Add Javadoc in PerformanceUtils.java

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Change scope of countVisitForms in SubmissionEventListener from public to private

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Make HTTP requests using the (now introduced) io.uhndata.cards.httprequests module which makes requests using the org.apache.httpcomponents module

This io.uhndata.cards.httprequests module will serve, in the future, as the interface through which CARDS will use to make all external HTTP requests.

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -116,10 +117,14 @@
+            importedAppointmentsCount += getUpcomingAppointments(token, this.daysToQuery, this.clinicNames[i]);
         }
+        // Update the performance counter
+        PerformanceUtils.updatePerformanceCounter(this.resolverFactory,
```
I might be overcautious here, but would it make sense to wrap any stats code in a `try ... catch (generic exception)` to make sure functionality is in no way halted by any random errors that may arise in the code for measuring performance?

marta- replied:
Nevermind, I see it's done properly inside `updatePerformanceCounter`.

marta- commented:
> Please note that the Imported Appointments performance counter increments each time the Torch import task runs and obtains upcoming visits and not when new visits are created or existing visits are modified. Should this behavior be changed?

No change, this is the right way afaict.

sdumitriu commented on modules/performance-notifications/src/main/java/io/uhndata/cards/performancenotifications/PerformanceUtils.java:
```diff
@@ -0,0 +1,146 @@
+        final String statName, final long incrementValue)
+    {
+        Map<String, Object> params = new HashMap<>();
+        params.put(ResourceResolverFactory.SUBSERVICE, "SlackNotifications");
```
`SlackNotifications` has no place in this module.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/submissioncounter/SubmissionEventListener.java:
```diff
@@ -0,0 +1,144 @@
+
+    private ResourceResolverFactory resolverFactory;
+    private ResourceResolver resolver;
+    private String submissionCounterName;
```
This is not used.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/slacknotifications/SlackNotificationsTask.java:
```diff
@@ -0,0 +1,169 @@
+    @Override
+    public void run()
+    {
+        LOGGER.warn("Running SlackNotificationsTask");
```
This should be  `debug` message.

```suggestion
        LOGGER.debug("Running SlackNotificationsTask");
```

sdumitriu commented on modules/performance-notifications/src/main/java/io/uhndata/cards/performancenotifications/PerformanceUtils.java:
```diff
@@ -0,0 +1,146 @@
+     * @param statName the name of the performance statistic to obtain its "today" and "total" values
+     * @return the map of 'today' and 'total' values for the performance statistic or null
+     */
+    public static Map<String, Long> getAndSetPreviousPerformanceStatistic(ResourceResolver resolver, String statName)
```
This very long name really bothers me. It's too verbose about its internals. How does `getAndReset` sound?

sdumitriu commented on modules/performance-notifications/src/main/java/io/uhndata/cards/performancenotifications/PerformanceUtils.java:
```diff
@@ -0,0 +1,146 @@
+     * @param statName the name of the performance statistic to query
+     * @return the map of 'today' and 'total' values for the performance statistic or null
+     */
+    public static Map<String, Long> getPerformanceStatistic(ResourceResolver resolver, String statName)
```
All the methods in here have the same redundant suffix. I would trim it:

```suggestion
    public static Map<String, Long> get(ResourceResolver resolver, String statName)
```

sdumitriu commented on modules/performance-notifications/src/main/java/io/uhndata/cards/performancenotifications/PerformanceUtils.java:
```diff
@@ -0,0 +1,146 @@
+     * @param statName the name of the performance statistic to increment
+     * @param incrementValue the value to increment the performance statistic by
+     */
+    public static void updatePerformanceCounter(final ResourceResolverFactory resolverFactory,
```
```suggestion
    public static void increment(final ResourceResolverFactory resolverFactory,
```

sdumitriu commented on modules/performance-notifications/src/main/java/io/uhndata/cards/performancenotifications/PerformanceUtils.java:
```diff
@@ -0,0 +1,146 @@
+        Map<String, Object> params = new HashMap<>();
+        params.put(ResourceResolverFactory.SUBSERVICE, "SlackNotifications");
+        try {
+            ResourceResolver resolver = resolverFactory.getServiceResourceResolver(params);
```
This should be a try-with-resources to simplify the code and make sure it is always closed.

sdumitriu commented on modules/performance-notifications/src/main/java/io/uhndata/cards/performancenotifications/PerformanceUtils.java:
```diff
@@ -0,0 +1,146 @@
+        return statsMap;
+    }
+
+    private static void incrementPerformanceStatistic(ResourceResolver resolver, String statName, long incrementBy)
```
```suggestion
    private static void increment(ResourceResolver resolver, String statName, long incrementBy)
```

sdumitriu reviewed (commented):
A few things that bother me:

1. The names of things. "Performance Statistics" isn't a very representative name, I would call these "Metrics".
    * the module name also adds `notifications`, but this module does not have any "notification" functionality, it just exposes counters, so I would name it `cards-metrics`
    * `PerformanceUtils` -> `Metrics`
 2. I'm not sure which code needs to have access to the repo: does each metrics user need to be given rights, or we assume that anyone can interface with the metrics, so the metrics module itself is responsible for getting its own session with the proper service user? On one hand, it's safer if we restrict who can access each counter, but on the other hand I would like to not have to pass resource resolvers on every call.
 3. Changing the name to `metrics` reminded me that there is already support for metrics in Sling, there's a dedicated sling-commons-metrics module that is already enabled and active, counting lots of things. Looking over it, I think it has almost all we need out of the box, except for a "getAndReset" functionality for which we can easily implement support, but with the major disadvantage that it does not support persistent counters: restarting the server will make every counter go back to 0. I think that's a blocker for us, right?

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Do not make reference to "SlackNotifications" in PerformanceUtils.java. Instead call it "PerformanceLogger".

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Remove unused private String submissionCounterName in SubmissionEventListener.java

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Rename PerformanceUtils.getAndSetPreviousPerformanceStatistic to PerformanceUtils.getAndReset

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Rename PerformanceUtils.getPerformanceStatistic to PerformanceUtils.get

sdumitriu commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -109,6 +109,7 @@
         "io.uhndata.cards.proms-backend:EmailNotifications=[sling-readall]",
         "io.uhndata.cards.proms-backend:UnsubmittedFormsRestriction=[sling-readall]",
         "io.uhndata.cards.proms-backend:TorchImporter=[proms-import-backend]",
+        "io.uhndata.cards.proms-backend:PerformanceLogger=[sling-performance-stats]",
```
This is the wrong place for this configuration (and the wrong module name). Since this is a core module, I would put the configuration in `distribution/src/main/features/core/cards.json`.

sdumitriu replied:
so, because `PerformanceUtils` is not an OSGi component, its own module is not taken into account when determining who is requesting a service session, so for now it's OK to leave the configuration here.

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Rename PerformanceUtils.updatePerformanceCounter to increment

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: In PerformanceUtils.java, rename private method incrementPerformanceStatistic to increment

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: In PerformanceUtils.java, get the Service Resource Resolver in a try-with-resources try block to ensure that it is always closed

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming apps, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: In SlackNotificationsTask.java, replace call to LOGGER.warn with a call to LOGGER.debug.

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming apps, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Rename "Performance Statistics" to "Metrics".

IntegralProgrammer committed:
CARDS-1492: Slack notifications: # of upcoming apps, # of initial emails sent, # of reminder emails sent, # of surveys completed

Address code review comment: Rename "Performance Statistics" to "Metrics".

IntegralProgrammer commented:
Addressed review comments and tested. Everything looks good - ready for re-review.

sdumitriu commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -73,7 +81,23 @@
+        "create path (sling:Folder) /Metrics/AppointmentSurveysSubmitted/total(nt:unstructured mixin mix:atomicCounter)",
+        "create path (sling:Folder) /Metrics/ImportedAppointments/prevTotal(nt:unstructured)",
+        "create path (sling:Folder) /Metrics/ImportedAppointments/total(nt:unstructured mixin mix:atomicCounter)",
+        "set properties on /Metrics/InitialEmailsSent/prevTotal \n   set value{Long} to 0 \n end",
+        "set properties on /Metrics/ReminderEmailsSent/prevTotal \n   set value{Long} to 0 \n end",
+        "set properties on /Metrics/TotalSurveysSubmitted/prevTotal \n   set value{Long} to 0 \n end",
+        "set properties on /Metrics/AppointmentSurveysSubmitted/prevTotal \n   set value{Long} to 0 \n end",
+        "set properties on /Metrics/ImportedAppointments/prevTotal \n   set value{Long} to 0 \n end"
```
```suggestion
        "set properties on /Metrics/InitialEmailsSent/prevTotal \n   default value{Long} to 0 \n end",
        "set properties on /Metrics/ReminderEmailsSent/prevTotal \n   default value{Long} to 0 \n end",
        "set properties on /Metrics/TotalSurveysSubmitted/prevTotal \n   default value{Long} to 0 \n end",
        "set properties on /Metrics/AppointmentSurveysSubmitted/prevTotal \n   default value{Long} to 0 \n end",
        "set properties on /Metrics/ImportedAppointments/prevTotal \n   default value{Long} to 0 \n end"
```

sdumitriu committed:
CARDS-1492: Slack notifications: # of upcoming appts, # of initial emails sent, # of reminder emails sent, # of surveys completed

Make sure the counters' previous values will not be reset to zero when restarting the server
